plugin/core/sessions.py

```diff
@@ -397,96 +483,93 @@ def _handle_request_workspace_configuration(self, params: Dict[str, Any], reques
-
-
-def create_session(config: ClientConfig,
-                   workspace_folders: List[WorkspaceFolder],
-                   env: dict,
-                   settings: Settings,
-                   on_pre_initialize: Optional[Callable[[Session], None]] = None,
-                   on_post_initialize: Optional[Callable[[Session], None]] = None,
-                   on_post_exit: Optional[Callable[[str], None]] = None,
-                   on_stderr_log: Optional[Callable[[str], None]] = None,
-                   bootstrap_client: Optional[Any] = None) -> Optional[Session]:
-
-    def with_client(client: Client) -> Session:
-        return Session(
-            config=config,
-            workspace_folders=workspace_folders,
-            client=client,
-            on_pre_initialize=on_pre_initialize,
-            on_post_initialize=on_post_initialize,
-            on_post_exit=on_post_exit)
-
-    session = None
-    if config.binary_args:
-        tcp_port = config.tcp_port
-        server_args = config.binary_args
-
-        if config.tcp_mode == "host":
-            socket = start_tcp_listener(tcp_port or 0)
-            tcp_port = socket.getsockname()[1]
-            server_args = list(s.replace("{port}", str(tcp_port)) for s in config.binary_args)
-
-        working_dir = workspace_folders[0].path if workspace_folders else None
-        process = start_server(server_args, working_dir, env, on_stderr_log)
-        if process:
-            if config.tcp_mode == "host":
-                client_socket, address = socket.accept()
-                transport = TCPTransport(client_socket)  # type: Transport
-                session = with_client(Client(transport, settings))
-            elif tcp_port:
-                transport = start_tcp_transport(tcp_port, config.tcp_host)
-                if transport:
-                    session = with_client(Client(transport, settings))
-                else:
-                    # try to terminate the process
-                    try:
-                        process.terminate()
-                    except Exception:
-                        pass
-            else:
-                session = with_client(attach_stdio_client(process, settings))
-    else:
-        if config.tcp_port:
-            transport = start_tcp_transport(config.tcp_port)
-            session = with_client(Client(transport, settings))
-        elif bootstrap_client:
-            session = with_client(bootstrap_client)
-        else:
-            debug("No way to start session")
-    return session
+        self.state = ClientStates.STOPPING
+        self.on_shutdown()
+        self.send_request(Request.shutdown(), self._handle_shutdown_result, self._handle_shutdown_result)
+
+    def _handle_shutdown_result(self, _: Any) -> None:
+        self.send_notification(Notification.exit())
+
+    def on_transport_close(self, exit_code: int, exception: Optional[Exception]) -> None:
+        super().on_transport_close(exit_code, exception)
```

Comment:
Again, don't know why are you calling `super()`. Must be missing something. :)

Reply:
The base class is `Client`.

Reply:
Wow. `Client` nust have looked like `object` to me when I looked at it. :)